libsolidity/interface/ABI.cpp

```diff
@@ -26,12 +26,31 @@ using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
+namespace
+{
+bool containsStoragePointer ( TypePointers const& _pointers )
+{
+	for (TypePointer const& pointer: _pointers)
+		if (pointer->dataStoredIn(DataLocation::Storage))
+			return true;
+
+	return false;
+}
+}
+
 Json::Value ABI::generate(ContractDefinition const& _contractDef)
 {
 	Json::Value abi(Json::arrayValue);
 
 	for (auto it: _contractDef.interfaceFunctions())
 	{
+		if (
+			_contractDef.isLibrary() &&
+			(it.second->stateMutability() > StateMutability::View ||
```

Comment:
What is the purpose of this condition?

Reply:
We want to exclude it if it is not `Pure` or `View`, at least according to your [comment](https://github.com/ethereum/solidity/issues/3409#issuecomment-360780792). 
The enum begins with `Pure, View, ... `

Reply:
Oh right! It would be good to get @axic's opinion about this if we really want it.

Reply:
From the chat:

axic: Reading through it, I think only pure functions should be exposed. Reading/modifying the state isn’t something what can be done via external interaction with the library, since the library doesn’t have its own state.

chris: I thought it might make sense to call library functions that then call other contract's view functions
ok, so perhaps put it differently: libraries were never meant to have a json-ABI at all
but people might use it, so I think we should include everything that is callable.

axic: Right, but we cannot enforce that wont try to read the state of the account of the library

chris: it's a tradeoff - you cannot read much state anyway, because everything that accepts a storage pointer will not be part of the json abi

axic: I’m not even sure exposing anything at all makes sense.
But exposing pure functions seems to be the safest.

chris: exactly, either not exposing anything or exposing the stuff that works